plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Completion/Completions.hs

```diff
@@ -0,0 +1,221 @@
+import           Data.Text.Utf16.Rope                          (Rope)
+import qualified Data.Text.Utf16.Rope                          as Rope
+import           Development.IDE                               as D
+import           Distribution.Compat.Lens                      ((^.))
+import           Ide.Plugin.Cabal.Completion.Types
+import qualified Language.LSP.Protocol.Lens                    as JL
+
+import           Ide.Plugin.Cabal.Completion.Completer.Simple
+import           Ide.Plugin.Cabal.Completion.Completer.Snippet
+import           Ide.Plugin.Cabal.Completion.Completer.Types   (Completer)
+import           Ide.Plugin.Cabal.Completion.Data
+import qualified Language.LSP.Protocol.Types                   as LSP
+import qualified Language.LSP.VFS                              as VFS
+import qualified System.FilePath                               as FP
+import           System.FilePath                               (takeBaseName)
+
+-- ----------------------------------------------------------------
+-- Public API for Completions
+-- ----------------------------------------------------------------
+
+{- | Takes information about the completion status within the file
+  and finds the correct completer to be applied.
+-}
+contextToCompleter :: Context -> Completer
+-- if we are in the top level of the cabal file and not in a keyword context,
+-- we can write any top level keywords or a stanza declaration
+contextToCompleter (TopLevel, None) =
+  snippetCompleter <> (constantCompleter $
+    Map.keys (cabalVersionKeyword <> cabalKeywords) ++ Map.keys stanzaKeywordMap)
+-- if we are in a keyword context in the top level,
+-- we look up that keyword in the top level context and can complete its possible values
+contextToCompleter (TopLevel, KeyWord kw) =
+  case Map.lookup kw (cabalVersionKeyword <> cabalKeywords) of
+    Nothing -> \recorder cData -> do
+      logWith recorder Warning $ LogUnknownKeyWordInContextError kw
+      noopCompleter recorder cData
+    Just l -> l
+-- if we are in a stanza and not in a keyword context,
+-- we can write any of the stanza's keywords or a stanza declaration
+contextToCompleter (Stanza s n, None) =
+  case Map.lookup s stanzaKeywordMap of
+    Nothing -> \recorder cData -> do
+      logWith recorder Warning $ LogUnknownStanzaNameInContextError s
+      noopCompleter recorder cData
+    Just l -> constantCompleter $ Map.keys (l n) ++ Map.keys stanzaKeywordMap
+-- if we are in a stanza's keyword's context we can complete possible values of that keyword
+contextToCompleter (Stanza s n, KeyWord kw) =
+  case Map.lookup s stanzaKeywordMap of
+    Nothing -> \recorder cData -> do
+      logWith recorder Warning $ LogUnknownStanzaNameInContextError s
+      noopCompleter recorder cData
+    Just m -> case Map.lookup kw (m n) of
+      Nothing -> \recorder cData -> do
+        logWith recorder Warning $ LogUnknownKeyWordInContextError kw
+        noopCompleter recorder cData
```

Comment:
To avoid manually wrapping `Completer`, how about a `loggedNoopCompleter (LogUnknownKeyWordInContextError kw)` that logs the Log message if invoked?